CIP-0030/README.md

```diff
@@ -233,13 +232,19 @@ Errors: `APIError`, `TxSignError`
 
 Requests that a user sign the unsigned portions of the supplied transaction. The wallet should ask the user for permission, and if given, try to sign the supplied body and return a signed transaction. If `partialSign` is true, the wallet only tries to sign what it can. If `partialSign` is false and the wallet could not sign the entire transaction, `TxSignError` shall be returned with the `ProofGeneration` code. Likewise if the user declined in either case it shall return the `UserDeclined` code. Only the portions of the witness set that were signed as a result of this call are returned to encourage dApps to verify the contents returned by this endpoint while building the final transaction.
 
-### api.signData(addr: cbor\<address>, sigStructure: cbor\<Sig_structure>): Promise\<Bytes>
+### api.signData(addr: cbor\<address>, payload: String): Promise\<cbor\<COSE_Sign1>>
 
 Errors: `APIError`, `DataSignError`
 
-This endpoint is due to be updated/finalized soon, see [discussion in the initial PR](https://github.com/cardano-foundation/CIPs/pull/88#issuecomment-954436243).
+This endpoint utilizes the [CIP-0008 signing spec](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0008/CIP-0008.md) for standardization/safety reasons. It allows the dApp to request the user to sign a payload conforming to said spec. The user's consent should be requested and the message to sign shown to the user. `payload` is hex-encoded bytes. The payment key from `addr` will be used for base, enterprise and pointer addresses to determine the EdDSA25519 key used. The staking will will be used for reward addresses. This key will be used to sign the COSE-Sign1 `Sig_structure` with the following headers set:
 
-This endpoint utilizes the [CIP-0008 signing spec](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0008/CIP-0008.md) for standardization/safety reasons. It allows the dApp to request the user to sign data conforming to said spec. The user's consent should be requested and the details of `sig_structure` shown to them in an informative way. The  Please refer to the CIP-0008 spec for details on how to construct the sig structure.
+* `alg` (1) - must be set to EdDSA (-8)
+* `kid` (4) - must be set to the Ed25519 public key bytes used to sign the `Sig_structure`
+* `"address"` - must be set to the raw binary bytes of the address as per the binary spec, without the CBOR binary wrapper tag
```

Comment:
@SebastienGllmt this is what I understood from the CIP-8 spec but it doesn't explicitly say which format to use.

here's the quote:
>For P1, the mapping of public keys to addresses has two main cases:
> 1. for Shelley addresses, one payment key maps to many addresses (base, pointer, enterprise)
> 2. for Byron addresses, chaincode and address attributes need to be combined with the public key to generate an address
>
> To resolve this, one SHOULD add the full address to the protected header when using a v2 address. The v2 addresses contain the chaincode + attributes and we can verify the address matches combining it with the verification public key.
>
> `? address: bstr,`

Reply:
What does "without the CBOR binary wrapper tag" mean? So in case of the serialization-lib, when I convert the addres to bytes, would that be correct or does that contain any CBOR tag?

Reply:
I wanted to write it without referring to any specific library. There are a couple crypto-related things in cardano-serialization-lib that have their `to_bytes()` have just some raw binary buffer for cases where people often are importing from non-CBOR contexts the bytes, whereas others will encode them in CBOR like BYTES(len) then the actual buffer. In the binary spec you have them all as just `bytes` which includes the tag though. I was actually mistaken for address and thought we did raw bytes but we do CBOR bytes (so 30 or 31 bytes with CBOR BYTES(29) tag before the 29 bytes of the address for enterprise for example). I think for hashes and keys we do raw bytes.

Reply:
I made an issue in the serialization lib to discus this Emurgo/cardano-serialization-lib#263

Reply:
> let someAddress = Address.from_bech32("addr1u8pcjgmx7962w6hey5hhsd502araxp26kdtgagakhaqtq8sxy9w7g");
protectedHeaders.set_header(
    Loader.Message.Label.new_text('address'),
    Loader.Message.CBORValue.new_bytes(someAddress.to_bytes())
);

How is this different? You use to_bytes() here also, which gives you the same result as when just taking the address directly in hex/bytes?

> it would be an issue (CIP-8/CIP-30-wise) as due to how CIP-30 defines cbor\<address>

I think many assume here it's simply the address in hex format the cardano ledger takes. If it's not truly cbor, then we should change the definition in CIP-30.

Reply:
>How is this different? You use to_bytes() here also, which gives you the same result as when just taking the address directly in hex/bytes?

I guess it's more of an issue _everywhere_ combining the serialization lib with things that take in `cbor<address>` as a parameter, since that would include the CBOR byte technically.

>If it's not truly cbor, then we should change the definition in CIP-30.

That would certainly make things easier. Address seems to be the only thing in the API where you would run into this as I don't see any hashes or keys in any of the parameters/returns...besides `submitTx` which is oddly defined as returning just `Promise<hash32>` instead of `Promise<cbor<hash32>>` but I don't think anyone noticed before.

So should we define those two types explicitly in the `Data Types` section as their own hex encoded strings and not bother with thinking of them as CBOR at all? We could still link to the spec's comments about the format of the address possibly or is it standard enough across the ecosystem to avoid all ambiguity without mentioning the byte format explicitly for addresses?

I guess we could also take in bech32 addresses as well but then having that in the type definition would make it harder for dApps to handle the addresses they get as API returns instead of being able to guarantee 1 single format.

Reply:
> So should we define those two types explicitly in the Data Types section as their own hex encoded strings and not bother with thinking of them as CBOR at all?

Yeah, that's a good idea. So far I was always looking at the specs when it comes to the types. For instance `api.getUsedAddresses` should also return `cbor<address>` according to the specs, however I'm returning them here in the raw format, which I assumed was the expected type. The cbor/bytes things is very confusing. I would really stick with the types of the cddl.

Reply:
Note that, addresses are one of the few things in the entire Cardano binary objects that are **NOT** CBOR-encoded (except Byron address, but let's ignore those for this specification -- although the spec should perhaps mention this decision!). 

One thing I don't quite understand here is: why even use bytes as an interface here? Shelley addresses are typically presented and stored as bech32 strings. While it is very much likely that a wallet implementing CIP-0030 would support bech32 encoding/decoding for its own sake, it's less likely that a dApp would. Or at least, requiring bytes here pushes that constraint on dApp developers. 

I'd argue that we should use plain string as bech32 for this interface (for address), and as stated above however, use an `ArrayBuffer` for the payload.

Reply:
> Note that, addresses are one of the few things in the entire Cardano binary objects that are **NOT** CBOR-encoded

Yes, was not precise enough. I meant according to the specs (CIP-0030) `cbor<address>` is expected, while in the actual alonzo specs of course it's not cbor, but just some raw bytes.

> One thing I don't quite understand here is: why even use bytes as an interface here? Shelley addresses are typically presented and stored as bech32 strings.

For consistency I would say. In CIP-0030 we return everything in bytes/hex. So when you grab the address there, you can directly insert it in that form into the `api.signData` endpoint.

> I'd argue that we should use plain string as bech32 for this interface (for address), and as stated above however, use an ArrayBuffer for the payload.

From an end user perspective it's better to have an encoding here like `utf8`, because it's better UX and the user can better verify the payload, since the wallet can display it. Or do you see use cases where you would want to have your own encoding?

Reply:
About addresses:

API consistency is better indeed so, if the rest already works with Base16 then let's stick to it. CBOR here is bonkers though. 

Alternatively, it wouldn't be much more complicated to support multiple encoding, defaulting to Base16 or Bech32, whatever makes the most sense. 


About the payload:

Well, the payload isn't necessarily a valid UTF-8 byte sequence, so it's only waiting for errors to use a plain string here.

Of course, wallets will want to present the information to users but I doubt they would present the full payload anyway since it's mostly gibberish from a user standpoint. Rather, wallets can and shall, whenever possible, extract information that are relevant to show for approval.

Reply:
> For consistency I would say. In CIP-0030 we return everything in bytes/hex. So when you grab the address there, you can directly insert it in that form into the api.signData endpoint.

What is the value of using CBOR elsewhere besides implicitly referring to ledger specs (and not having to update CIP when the specs are updated)? I think dApp developer UX is extremely important and it would be better with javascript object-based API.

Reply:
@KtorZ 

>API consistency is better indeed so, if the rest already works with Base16 then let's stick to it. CBOR here is bonkers though.

Yeah the CBOR thing was more of that is how it is technically defined right now by nature of referring to the ledger binary spec, but I agree it makes more sense to change to just raw bytes (encoded by hex).

>Well, the payload isn't necessarily a valid UTF-8 byte sequence, so it's only waiting for errors to use a plain string here.

I believe this was a mistake, it was meant to be a hex-encoded string type, to be consistent with CIP-0008 (message signing spec).

>Alternatively, it wouldn't be much more complicated to support multiple encoding, defaulting to Base16 or Bech32, whatever makes the most sense.

What about for returns? It's easier to accept multiple encodings in parameters but it becomes weird for return types. If you're always returning a specific one (bech or hex) so that users can easilly decode it, it would make usage with the other of the two less consistent imo.


@mkazlauskas 

>What is the value of using CBOR elsewhere

CBOR encoding is fairly ubiquitous in the Cardano ecosystem and it has support with various tooling already in a standardized way. If we had our own JSON encoding for everything here (which would amount to a lot of types once you think about how complciated some of these types are) you would still need to code conversions between that and CBOR (or worse, to other tool-specific encodings/constructions) in order to use other libraries together with CIP-30, unless you build some monolithic library that takes care of everything you could possibly need all within this CIP-30 specific format. The way it's defined here you can just call `YourLib.Transaction.fromCborBytes(bytes);` and everything is fine. I could see if there were some standardized JSON format for all these types that already existed, that we could use that instead. This was the reason we used cbor encoded bytes here at least despite the EIP-0012 (dapp connector spec for the Ergo blockchain) that this was initially based on we used JSON as that had a fairly standardized format there.

Reply:
> you would still need to code conversions between that and CBOR (or worse, to other tool-specific encodings/constructions) in order to use other libraries together with CIP-30

> The way it's defined here you can just call YourLib.Transaction.fromCborBytes(bytes);

Do you have any specific tools in mind? Using json schema for signTx/submitTx would remove dependency on `cardano-serialization-lib` for most dApps, which I think is great as it makes dApp development easier.

> I could see if there were some standardized JSON format for all these types that already existed, that we could use that instead.

cardano-js-sdk has [types](https://input-output-hk.github.io/cardano-js-sdk/modules/_cardano_sdk_core.Cardano.html) we could use as a reference. It also provides [utils](https://input-output-hk.github.io/cardano-js-sdk/modules/_cardano_sdk_core.coreToCsl.html) for type conversion between those types and `cardano-serialization-lib`

Reply:
> Using json schema for signTx/submitTx would remove dependency on cardano-serialization-lib for most dApps, which I think is great as it makes dApp development easier.

..and reduces the application's size by around 2MB